### DISCUSSION TITLE
sdcard library

### DISCUSSION THREAD

**Discussion by Strikelab** [micropython/micropython#12687] [RP2040 / Pico]
is there a simple way to find where you're hidding the sdcard library ?

**Comment by rkompass** [comment]
Yes: You google e.g. "sdcard lib micropython".
First finding is the doc on an perhaps present internal library, second a lib that you can put into your search path and import from there.

**Reply by Strikelab** [reply]
please don't say to me to google...
If that's your way to help ...
Don't help. 
Thanks

**Reply by andrewleech** [reply]
Well the answer by @rkompass was entirely accurate; as he stated the first two responses by Google are 100% relevant to your question and points directly to the source of the official sdcard driver.
https://github.com/micropython/micropython-lib/blob/master/micropython/drivers/storage/sdcard/sdcard.py

What he described is exactly how I would have found the same thing myself.

If you've got a reason you can't / won't accept Google based recommendations you should explain it?

**Reply by Strikelab** [reply]
I of course tried google searches before asking my question.
My response to @rkompass was not appropriate, he just wanted to help me by putting me on the path to the answer.
I finally found where this driver was.
Sorry for this misunderstanding, and thanks to @rkompass, and to you @andrewleech for your help.
Have a nice day.

**Reply by mendenm** [reply]
The sdcard driver on the micropython site is pretty weak.  I just spent this afternoon debugging and rewriting it.  Modern-ish SD cards didn't like the fact that many of the commands were sent without CRCs.  Also, there are some HUGE delays in the read and command login resulting in very slow transfers.  I will probably submit my modifications to the maintainer in the next couple days.  Here is my current work on it.  YMMV.
[sdcard.py.zip](https://github.com/micropython/micropython/files/13183364/sdcard.py.zip)

**Reply by rkompass** [reply]
@mendenm: Congratulations, that is good news. I hope you can do a pull request to micropython-lib for that.
Certainly a lot of background knowledge was enabling you to do that.
It would be interesting to learn something more about the new capabilities (what type of cards are read/writable now that weren't before, perhaps benchmarks regarding speed).

**Reply by mendenm** [reply]
before I start handing code back to any repo, I would appreciate some testing.  Here is my current draft of the sdcard.py and crc_7_16.py modules (which is needed by sdcard, but may end up being merged in).  The crc16 tries out to see if you are on a an arm thumb, and uses a very fast asm_thumb routine if it can, otherwise, both crc functions do use micropython.viper to be decent in speed.

If you initialize the sd card with use_crc = True, it will enable all crc checking, with a modest speed penalty.  I really implemented crcs as a check for other problems in the code, since I figured that if I could figure out the crcs, it wasn't completely broken.

[sdcard.20231029_1259.zip](https://github.com/micropython/micropython/files/13198668/sdcard.20231029_1259.zip)

**Comment by mendenm** [comment]
There was no deep knowledge.  I just happened to plug in a nice, new SDXC card, and it didn't work.  Some print statements showed that it failed as soon as incorrect (zero) CRCs started being sent, so I looked up the  CRC7 algorithm.

Speed is still slow, but for a lot of data-logging stuff, maybe OK.  At 120 MHz CPU, 40 MHz SPI on my Pi Pico, I get about 600kB/s write speed, and just over 1000 KB/s read speed.  Much of this may  due to small block sizes and poor buffer handling.

Note that before the changes, I was only getting 65 kB/s read speed, due to some time.sleep() in the read loop.

**Comment by rkompass** [comment]
@mendenm: I did a test of your driver.
Using the following program (modified for timing, original by @robert-hh), using a Blackpill with SPI2.
```py
# file write and erase test, includes timing
import machine, os, random, time, sdcard_m
from pyb import SPI, Pin
from time import ticks_ms, ticks_diff

spi = SPI(2, SPI.MASTER, baudrate=400_000, polarity=0, phase=0) # SCK=PB13, MISO=PB14, MOSI=PB15
sd = sdcard_m.SDCard(spi, Pin('PB12'), baudrate=8000000)
os.mount(sd, '/sd')
os.chdir('/sd')

def run(n=10, len=8000):
    t_wr = []
    t_rd = []
    bw = bytearray(len)
    br = bytearray(len)
    random.seed(0)
    for i in range (n):
        for j in range(len):
            bw[j] = random.getrandbits(8)
        name = "file" + str(i)
        print("create ", name)
        start = ticks_ms()
        f = open(name, "wb")
        f.write(bw)
        f.close()
        stop = ticks_ms()
        t_wr.append(ticks_diff(stop, start))  

    for e in os.ilistdir(""):
        print(e)

    random.seed(0)
    for i in range (n):
        for j in range(len):
            bw[j] = random.getrandbits(8)
        name = "file" + str(i)
        print("read ", name)
        start = ticks_ms()
        f = open(name, "rb")
        f.readinto(br)
        f.close()
        stop = ticks_ms()
        t_rd.append(ticks_diff(stop, start))  

        if bw != br:
            print("Read mismatch")
            input("Next")

    for i in range (n):
        name = "file" + str(i)
        print("delete ", name)
        os.remove(name)

    for e in os.ilistdir(""):
        print(e)

    print(t_wr)
    print(t_rd)
```
Results:
```
Filesystem tests with original Micropython-lib sdcard driver

Creating and writing files with length of 4000 bytes (ms):
[160, 157, 157, 159, 157, 157, 157, 159, 157, 182]
Reading of those files (ms):
[62, 63, 63, 63, 63, 63, 63, 63, 63, 63]

Creating and writing files with length of 8000 bytes (ms):
[206, 203, 204, 203, 203, 203, 203, 204, 203, 205]
Reading of those files (ms):
[111, 111, 111, 111, 111, 111, 111, 111, 111, 111]


Filesystem tests with improved sdcard driver

Creating and writing files with length of 8000 bytes (ms):
[188, 186, 189, 186, 186, 189, 186, 186, 189, 186]
Reading of those files (ms):
[100, 100, 100, 100, 100, 100, 100, 100, 100, 100]

Filesystem tests with improved sdcard driver, using optional baudrate=4000000 argument in sdcard init:

Creating and writing files with length of 8000 bytes (ms):
[70, 69, 73, 68, 71, 69, 68, 69, 70, 71]
Reading of those files (ms):
[29, 29, 30, 30, 29, 29, 29, 30, 30, 29]

Filesystem tests with improved sdcard driver, using optional baudrate=8000000 argument in sdcard init:

Creating and writing files with length of 8000 bytes (ms):
[51, 49, 53, 49, 71, 49, 49, 50, 52, 50]
Reading of those files (ms):
[18, 18, 18, 18, 18, 18, 18, 18, 18, 18]

Filesystem tests with original Micropython-lib sdcard driver, after adding an optional baudrate=8000000 argument in sdcard init:

Creating and writing files with length of 8000 bytes (ms):
[68, 390, 59, 59, 61, 60, 60, 60, 62, 59]
Reading of those files (ms):
[26, 26, 26, 26, 26, 26, 26, 26, 26, 26]
```
My observations from this so far are:
Your driver is faster. The biggest improvement is due to the ability to set a higher SPI baudrate at sdcard initialization.
Such an option is easily added to the original sdcard driver.
With higher SPI baudrate your driver excels at reducing file read times by another ca. 30%.

From a look at your code I see that in initialization there is a slightly better distinction between different card versions.
I did not test the crc option.

I think the code is worth to be looked after by the experts/maintainers.

Seeing the possible improvement by setting a higher SPI baudrate the question occurs: Which maximal baudrate is safe for a certain card type, and how can this type/baudrate be automatically determined?
You commented out the print of e.g. `[SDCard] v2 card`. My preference would rather be to have a more detailed log/print of the type of sdcard.

**Comment by mendenm** [comment]
good to know it is working out.  I have been doing some more general housekeeping on it to reduce code size and increase efficiency.  I have been doing all my testing at 24 MHz SPI speed, and even the very old cards I have seem OK with that.  I haven't looked into the process of probing cards to see what speed they actually support.

**Comment by rkompass** [comment]
O.k. I tried again with 25MHz, same tests:
original sdcard.py with baudrate=25_000_000 option:
write: [49, 48, 48, 51, 48, 48, 48, 50, 48, 48]
read: [25, 25, 25, 25, 25, 25, 25, 25, 25, 25]
your modified sdcard_m.py:
write: [35, 33, 34, 33, 32, 32, 36, 33, 32, 36]
read: [12, 12, 12, 12, 12, 12, 11, 12, 12, 12]

At higher SPI speed the speed advantage is even more pronounced.

How does e.g. Linux determine the SPI speed when accessing an sd card?
Can you tell the exact detail that gives this doubling of speed?

I should experiment with more different sd cards. You noted that the CRC option may allow to use sdcard types that otherwise are not accessible. Is that correct? Can we make this more specific?

On my Blackpill board I soldered Winbond W25Q128 (16 MByte) chip, together with capacitor, of course. See [here](https://github.com/mcauser/WEACT_F411CEU6).
It is connected to SPI1. The same test on /flash here gives 
write: [280, 274, 273, 283, 271, 276, 272, 274, 271, 269]
read: [3, 4, 4, 4, 3, 3, 3, 4, 4, 4]

Calculating these figures: 8000*8 bits / 3000 us = 21 bits/us => could be achieved with 24 MHz SPI speed (!?), although the SPI may be maximally set to 50 MHz (48 to be precise, as processor freq=96MHz). Theoretically dual mode SPI could be used too.

**Reply by rkompass** [reply]
Having the spi object I see that its speed (of SPI 2, used for the sdcard) is 12MHz. Setting to higher values apparently has no effect.
Found an old post stating (for Pyboard, which presumably applies for Blackpill board too) that
> SPI(1) is clocked from CPU/2; SPI(2) and SPI(3) are clocked from CPU/4.
> The SPI's can then have a prescaler (clock divider) between 2 and 256 (all powers of 2: 2 4 8 16 32 64 128 256).

So my above test was effectively done with 96MHz /4 / 2 = 12MHz. 
Read times reflect this: (3/2 times 8MHz SPI speed is compatible with 12ms = 2/3 times 18 ms read times).
One could try other boards for faster SPI perhaps.

**Comment by mendenm** [comment]
It was only the CRC on the commands that seemed to be necessary for one of my cards. The CRC is sent with commands independent of the CRC option; that only affects data transfers.  As of now, I have succeeded with three completely different cards (2 GB, 32 GB and 64 GB of different eras).

I may spend a little time looking at the SD card spec to see how to probe the maximum speed.   It might be useful to make this module really compliant.

Marcus

**Reply by rkompass** [reply]
One commented link in your `crc_7_16.py` again links again to this [spec](https://users.ece.utexas.edu/~valvano/EE345M/SD_Physical_Layer_Spec.pdf) , where at page 79 is stated about the CSD register splice [103:96] in CSD version 1 (same in CSD version 2):

TRAN_SPEED
The following table defines the maximum data transfer rate per one data line - TRAN_SPEED:
TRAN_SPEED
```
bit        code
2:0       transfer rate unit: 0=100kbit/s, 1=1Mbit/s, 2=10Mbit/s, 3=100Mbit/s, 4... 7=reserved
6:3       time value:    0=reserved, 1=1.0, 2=1.2, 3=1.3, 4=1.5, 5=2.0, 6=2.5, 7=3.0, 8=3.5, 9=4.0, A=4.5, B=5.0, C=5.5, D=6.0, E=7.0, F=8.0
7         reserved
```
Table 5-6: Maximum Data Transfer Rate Definition
Note that for current SD Memory Cards, this field shall be always 0_0110_010b (032h) which is equal to 25 MHz - the mandatory maximum operating frequency of SD Memory Card.
In High-Speed mode, this field shall be always 0_1011_010b (05Ah) which is equal to 50 MHz, and when the timing mode returns to the default by CMD6 or CMD0 command, its value will be 032h.

So it's easy to read this out and set accordingly. No need for baudrate at initialization.
You may insert 
```
        trsfr_unit = 10**((csd[12]&0b111)+4)
        trsfr_val = (5,10,12,13,15,20,25,30,35,40,45,50,55,60,70,80)[csd[12]>>3&0x0F]
        spi_freq = trsfr_unit*trsfr_val
        print('SPI freq:', spi_freq)
        self.init_spi(spi_freq)
```
at the end of init_card().

I noted that with the value mentioned in the above specification I get 25 Mhz, but my card (a SanDisc Ultra 8 GB) actually reports 10 MHz.

**Reply by rkompass** [reply]
Doing it accurately leads to a reduction of SPI freq to 6 MHz (integer SPI prescaler) and the read time now is 18 ms instead of 12 ms.
I would still opt for this, for safety.

**Comment by rkompass** [comment]
I found an error in `sdcard.py`:

After reading the specs I found that on line 103 it should be
`c_size_mult = ((csd[6] & 0b11) << 1) | csd[5] >> 7`
instead of 
`c_size_mult = ((csd[9] & 0b11) << 1) | csd[10] >> 7`
in v1 card version.

Do you have a v1 card version (< 2GB) perhaps? That should not be working or reporting the wrong size.

**Comment by mendenm** [comment]
I don't think I can actually find an old V1 card.  I will double check your correction against the specs, and put it in.
Thanks.

Marcus

**Reply by rkompass** [reply]
Found an old v1 card (64MB) and it's working.

**Correction:** There was no error: The CSD obviously is read in as bytes object indexed in reverse order:
So csd[15] translates to csd[0],  csd[14] to csd[1] and so on.
In the present code therefore there is no error.

The new SPI baudrate code has to be corrected for this though:
```py
        # set to high data rate now that it's initialised
        trsfr_unit = 10**((csd[3]&0b111)+4)
        trsfr_val = (5,10,12,13,15,20,25,30,35,40,45,50,55,60,70,80)[csd[3]>>3&0x0F]
        baudrate = trsfr_unit*trsfr_val
        print('SPI baudrate:', baudrate)
        self.init_spi(baudrate)
```
is the correct code: It gives 25 MHz even for my old 64 MB v1 sdcard, as expected and written in the specs.

> I noted that with the value mentioned in the above specification I get 25 Mhz, but my card (a SanDisc Ultra 8 GB) actually reports 10 MHz.

And 25MHz for this card now, as it should be. The read times are now at 12 ms too.

**Reply by rkompass** [reply]
More sdcards tested:

Transcend 4 GB: works,
Transcend 16 GB: works after trying to install it a second time, first time:
```
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "sdcard_m.py", line 59, in __init__
  File "sdcard_m.py", line 103, in init_card
  File "sdcard_m.py", line 165, in init_card_v2
OSError: timeout waiting for v2 card
```
SanDisk Extreme 256 GB: same timeout error, at every attempt to set up with `use_crc = False`.
Same error with `use_crc = True`.

BaseTech64 GB: Does not work the first time after power on:
```
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "sdcard_m.py", line 59, in __init__
  File "sdcard_m.py", line 101, in init_card
RuntimeError: SD card rejected CRC on command: 08
```
Second time it seems to work, regardless of `use_crc` option .
But the file write / read test reports once it starts to read:
```
read  file0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 39, in run
OSError: [Errno 2] ENOENT
```
This sdcard was full with a lot of different files.

All sdcards reported 25 MHz baudrate.

**Comment by mendenm** [comment]
I was just in the processing of annotating the CSD table, and making sure we were counting bytes in the right order.  

I am writing a function to extract bits from it by bit number, and if I use this, which will very slightly increase code size, it will greatly increase readability, and maybe allow future improvements which depend on specific bits from this block.

**Comment by mendenm** [comment]
The 25 MHz baud rate seems to be required by the standard for SPI.  I'm not sure any other value is actually permitted.

**Reply by rkompass** [reply]
> I was just in the processing of annotating the CSD table, and making sure we were counting bytes in the right order.

nice

> I am writing a function to extract bits from it by bit number, and if I use this, which will very slightly increase code size, it will greatly increase readability, and maybe allow future improvements which depend on specific bits from this block.

I'm not in support of this. At least not for a final version. The driver being small will be a criterion for acceptance in the end, probably.

**Reply by rkompass** [reply]
The above mentioned BaseTech64 GB card was write protected, couldn't unprotect or reformat it with Linux so far.

So from my assortment of cards I see no need of adding the CRC code. It seems necessary to check if the transmitted card data are correct, what may be assumed, and for changing e.g. CSD data, what we won't do. What about your cards?

It would be interesting to find out what prevented the 256 GB card from working.

The code for the SPI baudrate detection could be reduced to just check if the csd[3] is 0x32 => 25MHz or 0x5A => 50 MHz, which is apparently only the case if a high-speed mode was set, what this driver will not do.
Yet simpler: Just change the SPI baudrate to 25 MHz, which apparently fits all cards.

**Comment by mendenm** [comment]
I re-thought how to do it, and now it is much smaller, too.  I just convert the entire CSD 16-byte array to a big integer, and then directly extract bits.  Very clean, compact  and easy to read.

**Reply by mendenm** [reply]
here's the updated version, with much cleaner data field extraction.  I have verified that my 2 GB card is V1, so I have tested this on V1 and V2.  Let  me know what you think.
[sdcard.20231101_0604.zip](https://github.com/micropython/micropython/files/13225676/sdcard.20231101_0604.zip)

**Comment by mendenm** [comment]
OK, I did a bunch of cleanup, and have created my own fork of the micropython_lib repo, in eventual preparation for a pull request.

here's the updated version, with much cleaner data field extraction.  I have verified that my 2 GB card is V1, so I have tested this on V1 and V2.  Let  me know what you think. 

I also split up the files differently, so the crc7 code and crc16 code are in different modules.  This way, if you aren't using crc16 on the data, you don't have to have the crc16.py file at all.  It only gets imported when you turn on data crc.
[sdcard_20231101_1605.zip](https://github.com/micropython/micropython/files/13231344/sdcard_20231101_1605.zip)

**Comment by rkompass** [comment]
Hello @mendenm,

I looked at and briefly tested the last version.
From my certainly limited understanding / point of view I welcome your code as an interesting alternative/extended version of the sdcard code. Your changes to represent csd as a bigint may make the code clearer - at the cost of code size. To have the option to check the CRCs is a nice to have. At present from my tests I cannot see that it is a necessary option. My cards work well with the current `sdcard.py`. Only two of my sdards were not usable, neither by the standard sdcard.py nor by your driver: one of them is hopelessy write protected apparently (any help welcome to change that state) and the other 256GB card has exfat formatting, which is not usable by MP atm.
So a critical question is whether there are sdcards that are usable with your driver that are not usable with the present driver.
Provided that there are none: Following the logic of "if it's not broken, don't fix it" I would hesitate to recommend your new driver.
As soon as there were some sdcards found that would be enabled to work by your code improvements I would recommend the driver.

One argument for your driver is the read/write speed. This I found out depends heavily on the SPI baudrate. In my initial tests I used the default baudrate. Only later I noted that the current MP-lib version has an option for the baudrate. This makes sense as apparently all modern sdcards allow for 25 MHz to be used, but it may be desirable to use a slower rate for other reasons like protection from interference/crosstalk on longer SPI cables.
So  big improvements on read/write speed may be achieved with the standard sdcard.py driver by using the baudrate option for setting higher SPI speed.
Given that option is used up to the limit and one wants still higher read/write speed then your driver offers improvement: Somewhat faster write and up to 30-50% faster read speed. I wonder where this speed advantage actually originates from. If just a small part of changed code were responsible for that then I would vote for having a pull request for just this limited code change.

I think you should do a pull request on your sdcard driver version. You might get more feedback and as pull request it might be found by other developers more easily than as appended file in this conversation. Even if it is not accepted in mp-lib it may help/contribute to improve the driver in the future.
You might consider posting the code to [awesome micropython](https://awesome-micropython.com) perhaps.

**Comment by mendenm** [comment]
I think the bigint version is much smaller.  The calls to the bit extraction routine take almost no space at all.  I wii be posting another version sortly, with even more dead code removed; the separate v1 and v2 init routines are now merged, since there was little difference (the recheck for block size was redundant, wince that info also comes from the CSD). So, now I think the overall code size is much smaller.

The exfat formatting is, of course, a problem at a different layer of code.  A VFS for micropython to support exfat would have to be written.

I had one card which refused to take commands without the CRC7 correctly computed, so I think it is necessary.  Now that I have moved the CRC16 into a separate module, it doesn't cost any code size unless you use it.

I am still cleaning things up, but I think this will be a much more compact module, and it certainly reads and writes faster than the original.

Marcus

**Reply by rkompass** [reply]
> I had one card which refused to take commands without the CRC7 correctly computed, so I think it is necessary.

So, given this fact, I support the update of the driver. I hope that your pull request gets noted / accepted.
What kind of sdcard is this?

**Comment by mendenm** [comment]
@rkompass I have put an RFC on the micropython issues tracker, at:
https://github.com/micropython/micropython-lib/issues/763#issuecomment-1793425929
with a new update to the code.

**Reply by robert-hh** [reply]
You should make  a pull request to the lib instead providing the ZIP file.

**Comment by rkompass** [comment]
@mendenm: My above mentioned Basetech 64 GB sdcard, that was neither write- nor formattable has the PERM_WRITE_PROTECT bit set.
I took that sdcard from my old android device iirc. 
After adding
```py
print('sdcard perm_write_protect: {}'.format('on' if gb(csd_int, 13, 13) else 'off'))
print('sdcard tmp_write_protect: {}'.format('on' if gb(csd_int, 12, 12) else 'off'))
```
to the driver I got this message:
```
sdcard perm_write_protect: on
sdcard tmp_write_protect: off
```
Now I learnt that this bit is automatically set by the controller if the sdcard got too many write cycles (or perhaps got too "old" in the old-fashioned sense).
Also learnt that the TMP_WRITE_PROTECT bit may be set in the csd to make a read-only sdcard, that cannot be written or reformatted. See SDcard-Locker [here](https://github.com/vaginessa/SDCard-Locker-Attiny85/blob/master/sdlocker-tiny.cpp).

Apparently there is no sdcard lib in the Python/Micropython world that allows to solve/diagnose problems like that. At least I did an extensive search.
There is [libsd](https://github.com/wookey-project/libsd/tree/master), the above sdlocker and [f3](https://github.com/AltraMayor/f3) for C. Someone should port the essence of those and make an sdcard_util package for Python/Micropython.

**Comment by mendenm** [comment]
I will make a pull request.  I was hoping to get some people previewing the zips, before I started the more formal PR process.

I do have my fork, in which I am developing this, at 

https://github.com/mendenm/micropython-lib

but just grabbing the zip is easier for a tiny package.  Is posting the zips a no-no?

Marcus

**Reply by rkompass** [reply]
Perhaps the crc7 and crc16 may be made independent contributions to micropython-lib with user-defined polynomials (see discussion [12891](https://github.com/orgs/micropython/discussions/12891)).

**Reply by rkompass** [reply]
I just found that in the python docs:

> **binascii.crc_hqx(data, value)**
> Compute a 16-bit CRC value of data, starting with value as the initial CRC, and return the result. This uses the CRC-CCITT polynomial x16 + x12 + x5 + 1, often represented as 0x1021. This CRC is used in the binhex4 format.

This is the polynomial used by the sdcards, and that you @mendenm implemented in the fastest way (using table+asm_thumb) possible if one is not using C.
So the most natural approach would be to code a crc_hqx implementation as extension to the ubinascii module and adapt the sdcard driver to be able to use that (as an option). Should not be too hard to implement. After some discussion/thought if the table cannot be made smaller, e.g. having just 16 byte entries allowing for half the speed (processing 4 bits at a time) of the 512 byte table.

Hello @jimmo, would such an extension have a chance to be taken into the ubinascii module, given that that increases code size a bit?

**Comment by mendenm** [comment]
@rkompass @robert-hh 

In the sdcard.py update I am working on, there is (probably) legacy code which invokes a separate single-block read and single-block write command for those cases.  Do either of you think we should not just always use the multi-block command, and only get one block?  This removes quite a bit of code.  

Also, I changed the __init__ call to make it compatible with almost any future upgrades to the crc16 we do.  It now takes a function for the crc argument; if the function is None, no crc gets used. If a function is provided, it is used.  That  way the user can select  whatever optimized crc code is available.  If we upgrade the rp2 spi code to compute the crc by DMA, the spi.get_crc (whatever we end up calling it) can be passed in.

The current work I am doing on this is in: 

https://github.com/mendenm/micropython-lib/tree/sdcard-fixes/micropython/drivers/storage/sdcard 

I have slightly more cleanup before I do a PR on this.  I am pretty close.  I am in the process of converting the asserts into normal exceptions, and doing some other structural cleanup.  Those changes will appear in my repo later today.

**Reply by rkompass** [reply]
> It now takes a function for the crc argument; if the function is None, no crc gets used.

I think that's a good approach. Makes things simpler.

> Do either of you think we should not just always use the multi-block command, and only get one block? This removes quite a bit of code.

As non-expert at a first glance I cannot decide whether these are logically equivalent. They use different commands.
For legacy reasons the simple write could be a call to writeblocks, if this would work in a compatible way, what I don't now.

**Comment by mendenm** [comment]
@rkompass

Doing a multi-block read or write, and asking it to end after one block, should always be valid.  The multi-block commands are just 'at least one block, keep running until told to stop' and stopping after one block is valid.
